tests/python/frontend/pytorch/test_forward.py

```diff
@@ -0,0 +1,792 @@
+        baseline_input = baseline_input.cuda()
+    with torch.no_grad():
+        baseline_outputs = baseline_model(baseline_input)
+    if isinstance(baseline_outputs, tuple):
+        baseline_outputs = tuple(out.cpu().numpy() for out in baseline_outputs)
+    else:
+        baseline_outputs = (baseline_outputs.float().cpu().numpy(),)
+    output_shapes = [out.shape for out in baseline_outputs]
+    dtype = "float32"
+    input_name = "input0"
+    input_shapes = {input_name: list(baseline_input.shape)}
+    trace = torch.jit.trace(baseline_model, baseline_input).float().eval()
+    if torch.cuda.is_available():
+        trace = trace.cuda()
+    else:
+        trace = trace.cpu()
+
+    mod, params = relay.frontend.from_pytorch(trace, input_shapes)
+    compiled_input = {input_name: tvm.nd.array(baseline_input.cpu().numpy())}
+
+    with relay.build_config(opt_level=3):
+        for target, ctx in ctx_list():
+            relay_graph, relay_lib, relay_params = relay.build(mod, target=target, params=params)
+            relay_model = graph_runtime.create(relay_graph, relay_lib, ctx)
+            relay_model.set_input(**relay_params)
+            relay_model.set_input(**compiled_input)
+            relay_model.run()
+
+            for i, baseline_output in enumerate(baseline_outputs):
+                output_shape = baseline_output.shape
+                compiled_output = relay_model.get_output(
+                    i, tvm.nd.array(np.zeros(output_shape).astype(dtype), ctx)).asnumpy()
+
+                compiled_relay_output = relay_model.get_output(
+                    i, tvm.nd.array(np.zeros(output_shape).astype(dtype), ctx)).asnumpy()
+
+                assert_shapes_match(baseline_output, compiled_output)
+                tvm.testing.assert_allclose(baseline_output, compiled_output,
+                                            rtol=1e-3, atol=1e-3)
+
+                assert_shapes_match(baseline_output, compiled_relay_output)
+                tvm.testing.assert_allclose(baseline_output, compiled_relay_output,
+                                            rtol=1e-3, atol=1e-3)
+
+    del model_name
+    del baseline_model
+    torch.cuda.empty_cache()
+
+# Single operator tests
+def test_forward_add():
+    torch.set_grad_enabled(False)
+    input_shape = [10]
+
+    class Add1(Module):
+        def forward(self, *args):
+            return args[0] + args[0]
+
+    class Add2(Module):
+        def forward(self, *args):
+            return args[0] + 1
+
+    class Add3(Module):
+        def forward(self, *args):
+            ones = torch.ones(input_shape, dtype=torch.float)
+            if torch.cuda.is_available():
+                ones = ones.cuda()
+            return args[0] + ones
+
+    class Add4(Module):
+        def forward(self, *args):
+            ones = torch.ones(input_shape, dtype=torch.float)
+            if torch.cuda.is_available():
+                ones = ones.cuda()
+            return args[0] + ones
```

Comment:
Add3 and Add4 are the same.  Is it okay?

Reply:
Ah, let me look over the tests to clean it up. They used to use different input shapes before some refactoring.